ARROW-262: Start metadata specification document

wesm opened the pull request:
The purpose of this patch is to:

* Provide exposition and a place to clarify / provide examples illustrating the canonical metadata
* Begin providing definitions of logical types
* Where relevant, the data header metadata generated by a particular logical type (for example: strings produce one fewer buffer compared with List<UInt8> even though the effective memory layout is the same as a the nested type without any nulls in its child array)

This is not a complete specification and will require follow-up JIRAs to address more logical types and fill other gaps.

jacques-n commented on format/Metadata.md:
```diff
@@ -0,0 +1,238 @@
+
+## Record data headers
+
+A record batch is a collection of named, equal length Arrow arrays. One can be
```
Probably good to clarify the child fields are not equal length

wesm replied:
done

jacques-n commented on format/Metadata.md:
````diff
@@ -0,0 +1,238 @@
+}
+```
+
+The `nodes` and `buffers` fields are produced by a depth-first traversal /
````
Can we add an example of the traversal (i think we had one somewhere earlier)

jacques-n commented on format/Metadata.md:
```diff
@@ -0,0 +1,238 @@
+
+The integer endianness is currently set globally at the schema level. If a
+schema is set to be little-endian, then all integer types occurring within must
+be little-endian.
```
This should also clearly state that applies to offsets as well.

jacques-n commented on format/Metadata.md:
```diff
@@ -0,0 +1,238 @@
+type. The bits are numbered using least-significant bit (LSB) ordering.
+
+Like other fixed bit-width primitive types, boolean data appears as 2 buffers
+in the data header.
```
May be good to clarify that you mean that one is validity bitmap and value bitmap. (I think that is what you mean)

jacques-n commented on format/Metadata.md:
```diff
@@ -0,0 +1,238 @@
+The `List` logical type is the logical (and identically-named) counterpart to
+the List physical type.
+
+In data header form, the list field node contains 2 buffers:
```
This is a little bit confusing because lists also have child buffers. We probably also want to state that you can't have a list of no type (e.g. a list that doesn't have a child).  Actually, we may want to clarify whether that is allowed in the union and struct cases

julienledem replied:
Steven's PR (#99) is adding a Null type. would that mean we allow a list of nothing? (all values null)

wesm replied:
I'm adding a section for nested data to make the handling of child arrays more explicit

jacques-n commented on format/README.md:
```diff
@@ -7,6 +7,7 @@
 Currently, the Arrow specification consists of these pieces:
 
 - Physical memory layout specification (see Layout.md)
+- Metadata specification (see Metadata.md)
```
I'd start with the metadata concepts before getting into physical concepts.

julienledem commented on format/Metadata.md:
```diff
@@ -0,0 +1,238 @@
+
+The `nodes` and `buffers` fields are produced by a depth-first traversal /
+flattening of a schema (possibly containing nested types) for a given in-memory
+data set.
```
Is it important to flatten the representation of nodes and buffers here?
How about nesting the buffers inside FieldNode?
This would make the metadata more explicit and less error prone.
Right now the type in the fieldnode implies how many Buffers are found in the corresponding buffers list and what each one is.

julienledem replied:
discussed and resolved in https://github.com/apache/arrow/pull/124

julienledem commented on format/Metadata.md:
````diff
@@ -0,0 +1,238 @@
+hierarchy.
+
+```
+struct FieldNode {
````
struct can not be extended with new fields in the future.
Is it what we want? That means we can not add optional metadata in it later.

julienledem replied:
discussed and resolved in #124

julienledem commented on format/Metadata.md:
````diff
@@ -0,0 +1,238 @@
+For example, let's consider the schema
+
+```
+col1: Struct<a: Int32, b: List<Int64>, c: Float64>
````
is Struct what we call Tuple in Message.fbs ?

wesm replied:
`Struct` is a reserved word in some of the flatbuffers code generators

julienledem replied:
I think we should be consistent in the code.
Either we call it Struct_ (or something similar) in the fbs file or Tuple everywhere.

wesm replied:
I don't have an opinion here. I am OK with either Struct or Tuple. Thoughts?

wesm replied:
Seems like Struct is pretty consistently used for this concept in databases at least (in some databases, like BigQuery, the struct fields need not be named), so changing the metadata to `Struct_` seems reasonable

wesm replied:
Created ARROW-278 so this is not blocked

julienledem replied:
SGTM

julienledem commented on format/Metadata.md:
```diff
@@ -0,0 +1,238 @@
+
+We provide 3 types of floating point numbers as fixed bit-width primitive array
+
+- Half-float, 16-bit width
```
add it in Message.fbs?

wesm replied:
done

julienledem commented:
I think this is a very important document. Thanks for starting it. I made comments inline (both about the existing spec and the doc itself). This is a great start.

wesm commented on format/Metadata.md:
````diff
@@ -0,0 +1,238 @@
+```
+table Int {
+  bitWidth: int;
+  is_signed;
````
Our use of camelCase vs lower_with_underscores is inconsistent -- should we commit to one or the other (I'm more in the latter camp, but also that's my Python bias)?

wesm commented:
Rebased and responded to the outstanding comments (in the context of ongoing discussions in ARROW-264, ARROW-271). I can wait to make final updates here based on the outcomes of those patches if you'd like

wesm commented:
Rebased and resolved conflicts from ARROW-264. Reverted logical type name of string / utf8 to `Utf8` as it was originally (which failed the build; we can review names later)

wesm commented:
I created ARROW-278 as a follow up. There's a number of TODOs in the document that we will need to address in due course

julienledem commented:
+1
